File: DSL/DSL/Mobil.fs
namespace DSL

type Mobil = { Angle:float; Momentum:float; LastMove:float; X:float; Y:float }

module mobil =
    let toRadians angle = angle * 2.0 * System.Math.PI / 360.0
    let forward distance (mobil: Mobil) =
        let degrees =  mobil.Angle |> toRadians
        { mobil with
            X = mobil.X + distance * cos (degrees)
            Y = mobil.Y + distance * sin (degrees)
            LastMove = distance }

    let go (mobil:Mobil) = mobil |> forward (mobil.LastMove*mobil.Momentum)
    let turn angle (mobil: Mobil) =
        { mobil with
            Angle = (mobil.Angle + angle  ) % 360.0 }
    let momentum ponder (mobil: Mobil) =
        { mobil with
            Momentum = ponder }

File: DSL/DSL_Sprache/Mobil.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace DSL_Sprache
{
    class Mobil
    {
        public double AngleInRadians { get; set; }
        public double Momentum { get; set; }
        public double LastMove { get; set; }
        public double X { get; set; }
        public double Y { get; set; }

        private Point toPosition()
        {
            return new Point((int)X, (int)Y);
        }

        public IEnumerable<Point> PerformCommands(IEnumerable<Command> commands)
        {
            foreach (var command in commands)   
            {
                var forward = command as CommandForward;
                if (null != forward)
                {
                    yield return toPosition();
                    PerformCommandForward(forward.Distance);
                    yield return toPosition();
                    continue;
                }
                var turn = command as CommandTurn;
                if (null != turn)
                {
                    AngleInRadians += turn.AngleInDegrees * (Math.PI / 180d);
                    continue;
                }
                var momentum = command as CommandMomentum;
                if (null != momentum)
                {
                    Momentum = momentum.Ponder;
                    continue;
                }
                var go = command as CommandGo;
                if (null != go)
                {
                    yield return toPosition();
                    PerformCommandForward(LastMove * Momentum);
                    yield return toPosition();
                    continue;
                }
                var repeat = command as CommandRepeat;
                if (null != repeat)
                {
                    for (var i = 0; i < repeat.NumberOfTimes; i++)
                        foreach (var subc in PerformCommands(repeat.CommandsToRepeat))
                            yield return subc;
                    continue;
                }
                throw new ArgumentException("Command provided si not supported yet");
            }
        }

        private void PerformCommandForward(double distance)
        {
            X += distance * Math.Cos(AngleInRadians);
            Y += distance * Math.Sin(AngleInRadians);
            LastMove = distance;
        }
    }
}